File: turnstile_solver_service.go
package main

import (
	"bytes"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"time"
)

/*
Serviço de Resolução de Cloudflare Turnstile em Go
Usa Anti-Captcha API para resolver o captcha.

Endpoints:
  POST /solve       - Resolve um Turnstile e retorna o token
  POST /login       - Resolve Turnstile e faz login no 7k.bet.br
  GET  /health      - Verifica se o serviço está funcionando

Uso:
  go run turnstile_solver_service.go

Requisição:
  POST http://localhost:5099/solve
  {
    "sitekey": "0x4AAAAAAAykd8yJm3kQzNJc",
    "url": "https://7k.bet.br"
  }
*/

// Configuração
const (
	ANTICAPTCHA_KEY = "f80abc2cefe60bfec5c97f16294a1452"
	PROXY_HOST      = "pybpm-ins-hxqlzicm.pyproxy.io"
	PROXY_PORT      = "2510"
	PROXY_USER      = "liderbet1-zone-adam-region-br"
	PROXY_PASS      = "Aa10203040"
)

// Estruturas Anti-Captcha
type CreateTaskRequest struct {
	ClientKey string      `json:"clientKey"`
	Task      interface{} `json:"task"`
}

type TurnstileTask struct {
	Type       string `json:"type"`
	WebsiteURL string `json:"websiteURL"`
	WebsiteKey string `json:"websiteKey"`
}

type CreateTaskResponse struct {
	ErrorId          int    `json:"errorId"`
	ErrorCode        string `json:"errorCode"`
	ErrorDescription string `json:"errorDescription"`
	TaskId           int    `json:"taskId"`
}

type GetResultRequest struct {
	ClientKey string `json:"clientKey"`
	TaskId    int    `json:"taskId"`
}

type GetResultResponse struct {
	ErrorId  int    `json:"errorId"`
	Status   string `json:"status"`
	Solution struct {
		Token string `json:"token"`
	} `json:"solution"`
}

// Estruturas da API
type SolveRequest struct {
	SiteKey string `json:"sitekey"`
	URL     string `json:"url"`
	Timeout int    `json:"timeout"`
}

type SolveResponse struct {
	Success   bool    `json:"success"`
	Token     string  `json:"token,omitempty"`
	Error     string  `json:"error,omitempty"`
	TimeTaken float64 `json:"time_taken"`
}

type LoginRequest struct {
	Email    string `json:"email"`
	Password string `json:"password"`
}

type LoginResponse struct {
	Success      bool        `json:"success"`
	StatusCode   int         `json:"status_code"`
	Response     interface{} `json:"response"`
	Cookies      []string    `json:"cookies,omitempty"`
	TurnstileTime float64    `json:"turnstile_time"`
	Error        string      `json:"error,omitempty"`
}

// Cliente HTTP com proxy
func criarClienteComProxy() *http.Client {
	proxyURL, _ := url.Parse(fmt.Sprintf("http://%s:%s@%s:%s", PROXY_USER, PROXY_PASS, PROXY_HOST, PROXY_PORT))
	transport := &http.Transport{
		Proxy:           http.ProxyURL(proxyURL),
		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
	}
	return &http.Client{Transport: transport, Timeout: 60 * time.Second}
}

// Cliente HTTP sem proxy (para Anti-Captcha)
func criarClienteSemProxy() *http.Client {
	return &http.Client{Timeout: 120 * time.Second}
}

// Resolve Turnstile via Anti-Captcha
func resolverTurnstile(sitekey, pageURL string, timeout int) SolveResponse {
	start := time.Now()
	result := SolveResponse{Success: false}

	client := criarClienteSemProxy()

	// 1. Criar tarefa
	log.Printf("[Solver] Criando tarefa para %s...", pageURL)

	taskReq := CreateTaskRequest{
		ClientKey: ANTICAPTCHA_KEY,
		Task: TurnstileTask{
			Type:       "TurnstileTaskProxyless",
			WebsiteURL: pageURL,
			WebsiteKey: sitekey,
		},
	}

	jsonData, _ := json.Marshal(taskReq)
	resp, err := client.Post("https://api.anti-captcha.com/createTask", "application/json", bytes.NewBuffer(jsonData))
	if err != nil {
		result.Error = fmt.Sprintf("Erro ao criar tarefa: %v", err)
		return result
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	var taskResp CreateTaskResponse
	json.Unmarshal(body, &taskResp)

	if taskResp.ErrorId != 0 {
		result.Error = fmt.Sprintf("Anti-Captcha erro: %s - %s", taskResp.ErrorCode, taskResp.ErrorDescription)
		return result
	}

	log.Printf("[Solver] TaskId: %d", taskResp.TaskId)

	// 2. Aguardar resultado
	getResultReq := GetResultRequest{
		ClientKey: ANTICAPTCHA_KEY,
		TaskId:    taskResp.TaskId,
	}

	maxAttempts := timeout / 3
	if maxAttempts < 20 {
		maxAttempts = 20
	}

	for i := 0; i < maxAttempts; i++ {
		time.Sleep(3 * time.Second)

		jsonData, _ := json.Marshal(getResultReq)
		resp, err := client.Post("https://api.anti-captcha.com/getTaskResult", "application/json", bytes.NewBuffer(jsonData))
		if err != nil {
			continue
		}

		body, _ := io.ReadAll(resp.Body)
		resp.Body.Close()

		var getResult GetResultResponse
		json.Unmarshal(body, &getResult)

		if getResult.Status == "ready" {
			result.Success = true
			result.Token = getResult.Solution.Token
			result.TimeTaken = time.Since(start).Seconds()
			log.Printf("[Solver] Token obtido em %.2fs", result.TimeTaken)
			return result
		}

		if i%5 == 0 {
			log.Printf("[Solver] Aguardando... %ds", (i+1)*3)
		}
	}

	result.Error = "Timeout ao resolver Turnstile"
	result.TimeTaken = time.Since(start).Seconds()
	return result
}

// Faz login no 7k.bet.br
func fazerLogin7k(email, password, captchaToken string) LoginResponse {
	result := LoginResponse{Success: false}

	client := criarClienteComProxy()

	// 1. Primeiro acessa o site para pegar cookies do Cloudflare
	log.Println("[Login] Acessando site para obter cookies...")
	
	reqSite, _ := http.NewRequest("GET", "https://7k.bet.br", nil)
	reqSite.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36")
	reqSite.Header.Set("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8")
	reqSite.Header.Set("Accept-Language", "pt-BR,pt;q=0.9,en-US;q=0.8,en;q=0.7")
	
	respSite, err := client.Do(reqSite)
	if err != nil {
		log.Printf("[Login] Erro ao acessar site: %v", err)
	} else {
		respSite.Body.Close()
		log.Printf("[Login] Site acessado, cookies: %d", len(respSite.Cookies()))
	}

	// 2. Faz a requisição de login
	payload := map[string]string{
		"login":         email,
		"password":      password,
		"captcha_token": captchaToken,
	}

	jsonData, _ := json.Marshal(payload)
	log.Printf("[Login] Enviando login com token: %s...", captchaToken[:50])

	req, _ := http.NewRequest("POST", "https://7k.bet.br/api/auth/login", bytes.NewBuffer(jsonData))
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36")
	req.Header.Set("Origin", "https://7k.bet.br")
	req.Header.Set("Referer", "https://7k.bet.br/")
	req.Header.Set("Accept-Language", "pt-BR,pt;q=0.9,en-US;q=0.8,en;q=0.7")
	req.Header.Set("sec-ch-ua", `"Not_A Brand";v="8", "Chromium";v="120", "Google Chrome";v="120"`)
	req.Header.Set("sec-ch-ua-mobile", "?0")
	req.Header.Set("sec-ch-ua-platform", `"Windows"`)
	req.Header.Set("sec-fetch-dest", "empty")
	req.Header.Set("sec-fetch-mode", "cors")
	req.Header.Set("sec-fetch-site", "same-origin")

	// Adiciona cookies do site se existirem
	if respSite != nil {
		for _, cookie := range respSite.Cookies() {
			req.AddCookie(cookie)
		}
	}

	resp, err := client.Do(req)
	if err != nil {
		result.Error = fmt.Sprintf("Erro na requisição: %v", err)
		log.Printf("[Login] Erro: %v", err)
		return result
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	log.Printf("[Login] Status: %d, Body: %s", resp.StatusCode, string(body)[:min(200, len(body))])

	result.StatusCode = resp.StatusCode
	result.Success = resp.StatusCode == 200

	var jsonResp interface{}
	if err := json.Unmarshal(body, &jsonResp); err == nil {
		result.Response = jsonResp
	} else {
		result.Response = string(body)
	}

	// Captura cookies
	for _, cookie := range resp.Cookies() {
		result.Cookies = append(result.Cookies, fmt.Sprintf("%s=%s", cookie.Name, cookie.Value))
	}

	return result
}

// Handlers HTTP
func healthHandler(w http.ResponseWriter, r *http.Request) {
	json.NewEncoder(w).Encode(map[string]string{
		"status":  "ok",
		"service": "Turnstile Solver",
	})
}

func solveHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method != "POST" {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	var req SolveRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		json.NewEncoder(w).Encode(SolveResponse{
			Success: false,
			Error:   "JSON inválido",
		})
		return
	}

	if req.SiteKey == "" || req.URL == "" {
		json.NewEncoder(w).Encode(SolveResponse{
			Success: false,
			Error:   "sitekey e url são obrigatórios",
		})
		return
	}

	timeout := req.Timeout
	if timeout == 0 {
		timeout = 120
	}

	log.Printf("[API] Resolvendo Turnstile para %s", req.URL)

	result := resolverTurnstile(req.SiteKey, req.URL, timeout)

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(result)
}

func loginHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method != "POST" {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	var req LoginRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		json.NewEncoder(w).Encode(LoginResponse{
			Success: false,
			Error:   "JSON inválido",
		})
		return
	}

	if req.Email == "" || req.Password == "" {
		json.NewEncoder(w).Encode(LoginResponse{
			Success: false,
			Error:   "email e password são obrigatórios",
		})
		return
	}

	log.Printf("[API] Login para %s", req.Email)

	// 1. Resolver Turnstile
	solveResult := resolverTurnstile("0x4AAAAAAAykd8yJm3kQzNJc", "https://7k.bet.br", 120)

	if !solveResult.Success {
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(LoginResponse{
			Success:       false,
			Error:         "Falha ao resolver Turnstile: " + solveResult.Error,
			TurnstileTime: solveResult.TimeTaken,
		})
		return
	}

	// 2. Fazer login
	loginResult := fazerLogin7k(req.Email, req.Password, solveResult.Token)
	loginResult.TurnstileTime = solveResult.TimeTaken

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(loginResult)
}

func main() {
	fmt.Println("============================================================")
	fmt.Println("TURNSTILE SOLVER SERVICE (Go)")
	fmt.Println("============================================================")
	fmt.Println("Endpoints:")
	fmt.Println("  GET  /health  - Status do serviço")
	fmt.Println("  POST /solve   - Resolve um Turnstile")
	fmt.Println("  POST /login   - Resolve e faz login no 7k")
	fmt.Println("============================================================")
	fmt.Println("Iniciando em http://localhost:5099")
	fmt.Println("============================================================")

	http.HandleFunc("/health", healthHandler)
	http.HandleFunc("/solve", solveHandler)
	http.HandleFunc("/login", loginHandler)

	log.Fatal(http.ListenAndServe(":5099", nil))
}

File: turnstile_solver_service.py
"""
Serviço de Resolução de Cloudflare Turnstile
Usa Playwright para resolver automaticamente o captcha.

Endpoints:
  POST /solve - Resolve um Turnstile e retorna o token
  GET /health - Verifica se o serviço está funcionando

Uso:
  python turnstile_solver_service.py

Requisição:
  POST http://localhost:5099/solve
  {
    "sitekey": "0x4AAAAAAAykd8yJm3kQzNJc",
    "url": "https://7k.bet.br"
  }

Resposta:
  {
    "success": true,
    "token": "0.AbCdEf123...",
    "time_taken": 5.2
  }
"""

import sys
import io
sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')

import time
import json
import threading
from flask import Flask, request, jsonify
from playwright.sync_api import sync_playwright

app = Flask(__name__)

# Configuracao do Proxy
PROXY_CONFIG = {
    "server": "http://pybpm-ins-hxqlzicm.pyproxy.io:2510",
    "username": "liderbet1-zone-adam-region-br",
    "password": "Aa10203040"
}

# Pool de browsers para reutilizacao
browser_pool = []
pool_lock = threading.Lock()


def criar_pagina_turnstile(sitekey: str, page_url: str) -> str:
    """
    Cria uma pagina HTML com o Turnstile embutido para resolver.
    """
    return f'''
    <!DOCTYPE html>
    <html>
    <head>
        <title>Turnstile Solver</title>
        <script src="https://challenges.cloudflare.com/turnstile/v0/api.js" async defer></script>
    </head>
    <body>
        <h1>Resolvendo Turnstile...</h1>
        <div id="status">Aguardando...</div>
        
        <!-- Widget do Turnstile -->
        <div class="cf-turnstile" 
             data-sitekey="{sitekey}"
             data-callback="onSuccess"
             data-error-callback="onError">
        </div>
        
        <div id="token" style="display:none;"></div>
        
        <script>
            function onSuccess(token) {{
                document.getElementById('status').textContent = 'Resolvido!';
                document.getElementById('token').textContent = token;
                document.getElementById('token').style.display = 'block';
                console.log('TURNSTILE_TOKEN:' + token);
            }}
            
            function onError(error) {{
                document.getElementById('status').textContent = 'Erro: ' + error;
                console.log('TURNSTILE_ERROR:' + error);
            }}
        </script>
    </body>
    </html>
    '''


def resolver_turnstile_navegador(sitekey: str, page_url: str, timeout: int = 60) -> dict:
    """
    Resolve o Turnstile usando um navegador real.
    """
    start_time = time.time()
    result = {
        "success": False,
        "token": None,
        "error": None,
        "time_taken": 0
    }
    
    with sync_playwright() as p:
        # Lanca navegador com configuracoes anti-deteccao
        browser = p.chromium.launch(
            headless=True,
            args=[
                '--disable-blink-features=AutomationControlled',
                '--no-sandbox',
                '--disable-dev-shm-usage',
            ]
        )
        
        context = browser.new_context(
            proxy=PROXY_CONFIG,
            user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
            viewport={"width": 1366, "height": 768},
            locale="pt-BR"
        )
        
        page = context.new_page()
        
        # Variavel para capturar o token
        token_capturado = {"token": None}
        
        # Listener para capturar o token do console
        def on_console(msg):
            text = msg.text
            if text.startswith("TURNSTILE_TOKEN:"):
                token_capturado["token"] = text.replace("TURNSTILE_TOKEN:", "")
            elif text.startswith("TURNSTILE_ERROR:"):
                result["error"] = text.replace("TURNSTILE_ERROR:", "")
        
        page.on("console", on_console)
        
        try:
            # Metodo 1: Acessar diretamente a pagina do site
            print(f"[Solver] Acessando {page_url}...")
            page.goto(page_url, timeout=30000, wait_until="domcontentloaded")
            
            # Aguarda o Turnstile carregar e resolver
            print("[Solver] Aguardando Turnstile...")
            
            for i in range(timeout):
                time.sleep(1)
                
                # Tenta capturar o token do input hidden do Turnstile
                try:
                    # Busca o input com o token
                    token = page.evaluate("""
                        () => {
                            // Tenta varios seletores possiveis
                            const selectors = [
                                'input[name="cf-turnstile-response"]',
                                'input[name="cf_turnstile_response"]',
                                '[data-turnstile-response]',
                                '.cf-turnstile input[type="hidden"]'
                            ];
                            
                            for (const sel of selectors) {
                                const el = document.querySelector(sel);
                                if (el && el.value) {
                                    return el.value;
                                }
                            }
                            
                            // Tenta pegar do window
                            if (window.turnstile && window.turnstile.getResponse) {
                                return window.turnstile.getResponse();
                            }
                            
                            return null;
                        }
                    """)
                    
                    if token:
                        result["success"] = True
                        result["token"] = token
                        result["time_taken"] = round(time.time() - start_time, 2)
                        print(f"[Solver] Token capturado em {result['time_taken']}s")
                        break
                        
                except Exception as e:
                    pass
                
                # Verifica token do console
                if token_capturado["token"]:
                    result["success"] = True
                    result["token"] = token_capturado["token"]
                    result["time_taken"] = round(time.time() - start_time, 2)
                    print(f"[Solver] Token do console em {result['time_taken']}s")
                    break
                
                if i % 10 == 0:
                    print(f"[Solver] Aguardando... {i}s")
            
            if not result["success"]:
                result["error"] = "Timeout ao resolver Turnstile"
                
        except Exception as e:
            result["error"] = str(e)
            print(f"[Solver] Erro: {e}")
        
        finally:
            browser.close()
    
    return result


def resolver_turnstile_via_api(sitekey: str, page_url: str) -> dict:
    """
    Metodo alternativo: Resolve acessando a API do site diretamente.
    Alguns sites aceitam requests sem captcha se vier de um navegador valido.
    """
    start_time = time.time()
    result = {
        "success": False,
        "token": None,
        "error": None,
        "time_taken": 0
    }
    
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        
        context = browser.new_context(
            proxy=PROXY_CONFIG,
            user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
        )
        
        page = context.new_page()
        
        try:
            # Acessa o site para pegar cookies
            page.goto(page_url, timeout=30000)
            time.sleep(3)
            
            # Executa JavaScript para obter token do Turnstile
            token = page.evaluate(f"""
                async () => {{
                    return new Promise((resolve, reject) => {{
                        // Carrega o script do Turnstile
                        const script = document.createElement('script');
                        script.src = 'https://challenges.cloudflare.com/turnstile/v0/api.js';
                        script.onload = () => {{
                            // Renderiza o widget
                            turnstile.render('#turnstile-container', {{
                                sitekey: '{sitekey}',
                                callback: (token) => resolve(token),
                                'error-callback': (err) => reject(err)
                            }});
                        }};
                        
                        // Cria container
                        const container = document.createElement('div');
                        container.id = 'turnstile-container';
                        document.body.appendChild(container);
                        document.head.appendChild(script);
                        
                        // Timeout
                        setTimeout(() => reject('Timeout'), 60000);
                    }});
                }}
            """)
            
            if token:
                result["success"] = True
                result["token"] = token
                result["time_taken"] = round(time.time() - start_time, 2)
                
        except Exception as e:
            result["error"] = str(e)
        
        finally:
            browser.close()
    
    return result


# ============================================================
# ENDPOINTS DA API
# ============================================================

@app.route("/health", methods=["GET"])
def health():
    return jsonify({"status": "ok", "service": "Turnstile Solver"})


@app.route("/solve", methods=["POST"])
def solve():
    """
    Resolve um Cloudflare Turnstile.
    
    Request:
        {
            "sitekey": "0x4AAAAAAAykd8yJm3kQzNJc",
            "url": "https://7k.bet.br",
            "timeout": 60  (opcional)
        }
    
    Response:
        {
            "success": true,
            "token": "0.AbCdEf123...",
            "time_taken": 5.2
        }
    """
    try:
        data = request.get_json()
        
        sitekey = data.get("sitekey")
        page_url = data.get("url")
        timeout = data.get("timeout", 60)
        
        if not sitekey or not page_url:
            return jsonify({
                "success": False,
                "error": "sitekey e url sao obrigatorios"
            }), 400
        
        print(f"\n[API] Resolvendo Turnstile para {page_url}")
        print(f"[API] SiteKey: {sitekey}")
        
        result = resolver_turnstile_navegador(sitekey, page_url, timeout)
        
        if result["success"]:
            return jsonify(result)
        else:
            return jsonify(result), 500
            
    except Exception as e:
        return jsonify({
            "success": False,
            "error": str(e)
        }), 500


@app.route("/solve-and-login", methods=["POST"])
def solve_and_login():
    """
    Resolve o Turnstile e faz login no 7k.bet.br.
    
    Request:
        {
            "email": "user@email.com",
            "password": "senha123"
        }
    
    Response:
        {
            "success": true,
            "user": {...},
            "cookies": [...]
        }
    """
    try:
        data = request.get_json()
        email = data.get("email")
        password = data.get("password")
        
        if not email or not password:
            return jsonify({
                "success": False,
                "error": "email e password sao obrigatorios"
            }), 400
        
        print(f"\n[API] Login para {email}")
        
        # Resolve o Turnstile
        turnstile_result = resolver_turnstile_navegador(
            "0x4AAAAAAAykd8yJm3kQzNJc",
            "https://7k.bet.br",
            60
        )
        
        if not turnstile_result["success"]:
            return jsonify({
                "success": False,
                "error": "Falha ao resolver Turnstile",
                "details": turnstile_result["error"]
            }), 500
        
        # Faz login com o token
        import requests as req
        
        login_resp = req.post(
            "https://7k.bet.br/api/auth/login",
            json={
                "login": email,
                "password": password,
                "captcha_token": turnstile_result["token"]
            },
            headers={
                "Content-Type": "application/json",
                "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
                "Origin": "https://7k.bet.br",
                "Referer": "https://7k.bet.br/"
            },
            proxies={
                "http": f"http://{PROXY_CONFIG['username']}:{PROXY_CONFIG['password']}@pybpm-ins-hxqlzicm.pyproxy.io:2510",
                "https": f"http://{PROXY_CONFIG['username']}:{PROXY_CONFIG['password']}@pybpm-ins-hxqlzicm.pyproxy.io:2510"
            },
            timeout=30
        )
        
        return jsonify({
            "success": login_resp.status_code == 200,
            "status_code": login_resp.status_code,
            "response": login_resp.json() if login_resp.headers.get("content-type", "").startswith("application/json") else login_resp.text,
            "cookies": dict(login_resp.cookies),
            "turnstile_time": turnstile_result["time_taken"]
        })
        
    except Exception as e:
        return jsonify({
            "success": False,
            "error": str(e)
        }), 500


if __name__ == "__main__":
    print("=" * 60)
    print("TURNSTILE SOLVER SERVICE")
    print("=" * 60)
    print("Endpoints:")
    print("  GET  /health        - Status do servico")
    print("  POST /solve         - Resolve um Turnstile")
    print("  POST /solve-and-login - Resolve e faz login no 7k")
    print("=" * 60)
    print("Iniciando em http://localhost:5099")
    print("=" * 60)
    
    app.run(host="0.0.0.0", port=5099, debug=False, threaded=True)

